pkg/alertmanager/multitenant_test.go

```diff
@@ -1202,17 +1212,26 @@ func TestAlertmanager_StateReplicationWithSharding(t *testing.T) {
 				require.Regexp(t, regexp.MustCompile(`{"silenceID":".+"}`), string(body))
 			}
 
-			metrics := registries.BuildMetricFamiliesPerUser()
-
 			// If sharding is not enabled, we never propagate any messages amongst replicas in this way, and we can stop here.
 			if !tt.withSharding {
+				metrics := registries.BuildMetricFamiliesPerUser()
+
 				assert.Equal(t, float64(1), metrics.GetSumOfGauges("cortex_alertmanager_silences"))
 				assert.Equal(t, float64(0), metrics.GetSumOfCounters("cortex_alertmanager_state_replication_total"))
 				assert.Equal(t, float64(0), metrics.GetSumOfCounters("cortex_alertmanager_state_replication_failed_total"))
 				return
 			}
 
+			var metrics util.MetricFamiliesPerUser
+
 			// 5. Then, make sure it is propagated successfully.
+			//    Replication is asynchronous, so we may have to wait a short period of time.
+			assert.Eventually(t, func() bool {
+				metrics = registries.BuildMetricFamiliesPerUser()
+				return (float64(tt.replicationFactor) == metrics.GetSumOfGauges("cortex_alertmanager_silences") &&
+					float64(tt.replicationFactor) == metrics.GetSumOfCounters("cortex_alertmanager_state_replication_total"))
+			}, 5*time.Second, 100*time.Millisecond)
+
 			assert.Equal(t, float64(tt.replicationFactor), metrics.GetSumOfGauges("cortex_alertmanager_silences"))
 			assert.Equal(t, float64(tt.replicationFactor), metrics.GetSumOfCounters("cortex_alertmanager_state_replication_total"))
```

Comment:
[nit] This is already checked by the `assert.Eventually()` above.

Reply:
I left these in intentionally so the `assert.Equal` prints out the value of the counter (and expected value), but I agree it's redundant.